File: javascript/src/stack.ts
interface IStack<T> {
  push(value: T): void;
  pop(): T | undefined;
  peek(): T | undefined;
  size(): number;
  toString(formatter: (value: T) => string): string;
}

// JS is pretty easy since arrays have built in stack methods
export class Stack<T> implements IStack<T> {
  private values: T[];

  constructor(arr?: T[]) {
    if (arr) {
      this.values = [...arr];
    } else {
      this.values = [];
    }
  }

  push(value: T): void {
    this.values.push(value);
  }

  pop(): T | undefined {
    return this.values.pop();
  }

  peek(): T | undefined {
    return this.values.at(this.values.length - 1);
  }

  size(): number {
    return this.values.length;
  }

  toString(formatter: (value: T) => string): string {
    return "[" + this.values.map((v) => formatter(v)).join(",") + "]";
  }
}

File: java/src/main/java/com/example/Stack.java
package com.example;

import java.util.ArrayList;
import java.util.Optional;

import com.example.common.FormatterFunction;

interface IStack<T> {
    public void push(T value);

    public Optional<T> pop();

    public Optional<T> peek();

    public int size();

    public String toString(FormatterFunction<T> formatter);
}

public class Stack<T> implements IStack<T> {
    private ArrayList<T> values;

    public Stack() {
        this.values = new ArrayList<>();
    }

    public Stack(T[] arr) {
        this();
        for (T t : arr) {
            this.values.add(t);
        }
    }

    public void push(T value) {
        this.values.add(value);
    }

    public Optional<T> pop() {
        if (this.values.isEmpty()) {
            return Optional.empty();
        } else {
            return Optional.of(this.values.remove(this.values.size() - 1));
        }
    }

    public Optional<T> peek() {
        if (this.values.isEmpty()) {
            return Optional.empty();
        } else {
            return Optional.of(this.values.get(this.values.size() - 1));
        }
    }

    public int size() {
        return this.values.size();
    }

    public String toString(FormatterFunction<T> formatter) {
        ArrayList<String> elems = new ArrayList<>();
        for (T t : this.values) {
            elems.add(formatter.apply(t));
        }
        return "[" + String.join(",", elems) + "]";
    }
}
